core/src/util/pbfParser.cpp

```diff
@@ -139,102 +246,111 @@ Feature PbfParser::getFeature(ParserContext& _ctx, protobuf::message _featureIn)
-        if (tagValue >= 0) {
-            properties.emplace_back(_ctx.keys[tagKey], _ctx.values[tagValue]);
-        }
+    if (!extractTags(_ctx, tagsMsg)) {
+        return;
     }
-    feature.props.setSorted(std::move(properties));
 
-    switch(feature.geometryType) {
-        case GeometryType::points:
-            feature.points.insert(feature.points.begin(),
-                                  _ctx.geometry.coordinates.begin(),
-                                  _ctx.geometry.coordinates.end());
-            break;
+    if (!_sink.matchFeature(feature)) {
+        return;
+    }
 
-        case GeometryType::lines:
-        {
-            auto pos = _ctx.geometry.coordinates.begin();
-            for (int length : _ctx.geometry.sizes) {
+    extractGeometry(_ctx, geomMsg);
+
+    switch(feature.geometryType) {
+        case GeometryType::points: {
+                feature.points.clear();
+                feature.points.insert(feature.points.begin(),
+                                      _ctx.coordinates.begin(),
+                                      _ctx.coordinates.end());
+                break;
+        }
+        case GeometryType::lines: {
+            int offset = 0;
+            feature.lines.clear();
+            for (int length : _ctx.numCoordinates) {
                 if (length == 0) { continue; }
+
                 Line line;
                 line.reserve(length);
-                line.insert(line.begin(), pos, pos + length);
-                pos += length;
+
+                line.insert(line.begin(),
+                            _ctx.coordinates.begin() + offset,
+                            _ctx.coordinates.begin() + offset + length);
+
+                offset += length;
                 feature.lines.emplace_back(std::move(line));
             }
             break;
         }
-        case GeometryType::polygons:
-        {
-            auto pos = _ctx.geometry.coordinates.begin();
-            auto rpos = _ctx.geometry.coordinates.rend();
-            for (int length : _ctx.geometry.sizes) {
+        case GeometryType::polygons: {
+            int offset = 0;
+            WindingOrder windingOrder;
+
+            feature.polygons.clear();
+
+            for (int length : _ctx.numCoordinates) {
                 if (length == 0) { continue; }
-                float area = signedArea(pos, pos + length);
-                if (area == 0) {
-                    pos += length;
-                    rpos -= length;
+                auto it = _ctx.coordinates.begin() + offset;
+
+                double area = signedArea(it, it + length);
+                if (area == 0.f) {
+                    offset += length;
                     continue;
                 }
-                int winding = area > 0 ? 1 : -1;
-                // Determine exterior winding from first polygon.
-                if (_ctx.winding == 0) {
-                    _ctx.winding = winding;
+                WindingOrder winding(area);
```

Comment:
I agree that this would be part of a different PR. I'm not sure if it makes things simpler to use sign of area instead of winding. It's hard to not think of the default math coordinate system when talking about winding order. Whether one interprets the coordinates a y-down to visualize the geometry is not relevant to the algorithm working with them.  

IMO we just flip y-axis in too many places - and got confused by that :)